Python/initconfig.c

```diff
@@ -579,20 +568,75 @@ PyConfig_Clear(PyConfig *config)
 #undef CLEAR
 }
 
+// Future enhancement: consider making these public in patchlevel.h
+#define _Py_GET_MAJOR_MINOR(hexversion) ((hexversion & 0xFFFF0000) >> 16)
+#define _Py_GET_RELEASE_LEVEL(hexversion) ((hexversion & 0x000000F0) >> 4)
+#define _Py_HAS_NOMINALLY_FROZEN_ABI(level) ((level == 0xC) || (level == 0xF))
 
 PyStatus
-_PyConfig_InitCompatConfig(PyConfig *config)
+_Py_CheckVersionCompat(uint32_t header_version)
 {
-    size_t struct_size = config->struct_size;
-    memset(config, 0, sizeof(*config));
-    config->struct_size = struct_size;
+    uint32_t runtime_version = PY_VERSION_HEX;
+
+    // Exact matches are always fine
+    if (header_version == runtime_version) {
+        return _PyStatus_OK();
+    }
+
+    // Possible future enhancement: support string formatting for init errors
 
-    PyStatus status = config_check_struct_size(config);
+    // Only allow inexact matches for release candidates and final releases
+    uint8_t header_level = _Py_GET_RELEASE_LEVEL(header_version);
+    if (!_Py_HAS_NOMINALLY_FROZEN_ABI(header_level)) {
+        if (header_level != 0xA && header_level != 0xB) {
+            return _PyStatus_ERR(
+                "Embedding application failed to specify a valid header "
+                "release level"
+            );
+        }
+        return _PyStatus_ERR(
+            "Embedding applications built against a pre-release API version "
+            "must be linked to the exact corresponding library version"
+        );
+    }
+    uint8_t runtime_level = _Py_GET_RELEASE_LEVEL(runtime_version);
+    if (!_Py_HAS_NOMINALLY_FROZEN_ABI(header_level)) {
+        return _PyStatus_ERR(
+            "Embedding applications linked against a pre-release library "
+            "must be built against the exact corresponding header version"
+        );
+    }
+
+    // Otherwise require that the X.Y part of the version match
+    if (_Py_GET_MAJOR_MINOR(header_level) != _Py_GET_MAJOR_MINOR(runtime_level)) {
+        return _PyStatus_ERR(
+            "Embedding applications built against the headers for Python X.Y.z "
+            "must be linked against a library in the same Python X.Y series"
+        );
+    }
+
+    return _PyStatus_OK();
+}
+#undef _Py_GET_MAJOR_MINOR
+#undef _Py_GET_RELEASE_LEVEL
+#undef _Py_HAS_FROZEN_RUNTIME_API
+
+PyStatus
+_PyConfig_InitCompatConfig(PyConfig *config)
+{
+    uint32_t header_version = config->header_version;
+    PyStatus status = _Py_CheckVersionCompat(header_version);
     if (_PyStatus_EXCEPTION(status)) {
         _PyStatus_UPDATE_FUNC(status);
         return status;
     }
 
+    // Future enhancement: allow old config struct layouts if an old
+    // header_version is specified
+    size_t struct_size = sizeof(*config);
+    memset(config, 0, struct_size);
+    config->header_version = header_version;
```

Comment:
Note that prior to this change, the old code *didn't* pay attention to the `struct_size` setting (it wasn't used in the `memset` call), and if it had been naively adjusted to do so, could have resulted in clearing an arbitrary amount of memory if the `struct_size` field hadn't been initialised correctly by the caller.

With an explicit API/ABI version passed in, we'd never be trusting the supplied value directly - we'd be using it to select amongst a predefined set of known possible structure sizes.

Reply:
Just noting that the previous code *didn't* naively trust the supplied value: anything that didn't match the expected value was reported as a size mismatch error.